Module initializers spec proposal (#35140)
dotnet/runtime: docs/design/specs/Ecma-335-Augments.md

2) 0.5 GB is a very large heap. Having such a big PE file seems unreasonable and very rare scenario (if it exists at all).

3) Having 3 spare bits available is very beneficial for the implementation. It allows to represent WinRT projected strings, namespaces, etc. in very efficient way. If we had to represent heap indices with all 32 bits it would bloat various structures and increase memory pressure. PE files over 0.5 GB of size are very rare, but the overhead would affect all compilers and tools working with the metadata reader.

## Module Initializer

> All modules may have a module initializer. This method shall be static, a member of the module, take no parameters, return no value, be marked with **rtspecialname** and **specialname**, and be named `.cctor`.

Reviewer: ```suggestion
All modules may have a module initializer. This method shall be a global method, static, take no parameters, return no value, be marked with **rtspecialname** and **specialname**, and be named `.cctor`.
```

The ECMA spec does not explain "member of the module" anywhere. It does explain global methods.

Reviewer 2: It does however refer to the special `<Module>` class. Describing the module cctor as being a member of the `<Module>` class would also be correct. However, I prefer Jan's formulation, as global methods are a concept in the standard, which is easier to understand. Also, global methods are defined as always being static, so it is redundant, although acceptable to describe the .cctor as a static global method.

There are no limitations on what code is permitted in a module initializer. Module initializers are permitted to run and call both managed and unmanaged code.

### Module Initialization Guarantees

The CLI shall provide the following guarantees regarding module initialization:

> 1. A module initializer is executed at, or sometime before, first access to any types, methods, or data defined in the module.

Reviewer: ```suggestion
1. A module initializer is executed at, or sometime before, first access to any static field or first invocation of any method defined in the module.
```

2. A module initializer shall run exactly once for any given module unless explicitly called by user code.

> 3. No method other than those called directly or indirectly from the module initializer will be able to access the types, methods, or data in a module before its initializer completes execution.

Author: 1 and 3 *sound like* they would include access via reflection. However, `Assembly.GetTypes` currently returns type instances without triggering the module initializer of the module that the types came from. This is troublesome for plugin situations that need to hook AssemblyResolve/AssemblyLoadContext.Resolving before any of the types get a chance to load, as in the case of VSTest adapters.

1. Can the runtime be fixed?

2. Should reflection be explicitly mentioned here?

Reviewer: This is using similar language regular constructors (e.g. look for "first access to any static field").

Acquiring reflection object does not count as trigger for regular constructors, so I do not think it should count as trigger for module constructors either. Fixing it would be a potential breaking change - the code out there may be depending on the fact that loading an assembly via Assembly.Load does not actually run any code in it.

Author: Thanks. Would you consider reflection to be something worth mentioning at all, or is it too obvious if you know the context?

Reviewer: I do not think this needs to mention reflection.

> 3. No method other than those called directly or indirectly from the module initializer will be able to access the types, methods, or data in a module before its initializer completes execution.

Reviewer 3: Do we need to describe what happens when there are races and deadlocks between module initializers à la II.10.5.3.3 of the ECMA-335 spec?

Also would an initializer that throws an exception be considered as having "completed execution"?

Reviewer: It should just say that it is same as for type initializers. Something like:

The module initializers should be treated as if it were type initializers of type `<Module>` for the purpose of guarantees about races and deadlocks (II.10.5.3.3).

Author: @jkotas Do I understand correctly that module initializers *are* type initializers of the `<Module>` type?

Author: Or:

> As the type initializer for the `<Module>` type, a module initializer has the same race and deadlock guarantees (II.10.5.3.3) as any other type initializer.

Reviewer: Module initializers behave a lot like initializers of the `<Module>` type. 

If we were to say that module initializer is the type initializer of the <Module> type, I think it should be earlier in the chapter and the module initializer guarantees should be described as additional guarantees on top of the type initializer guarantees.

Author: Does this definition allow for the possibility that a module initializer is not a type initializer of the \<Module> type?

> This method shall be a global method, static, take no parameters, return no value, be marked with **rtspecialname** and **specialname**, and be named `.cctor`.

Author: > Module initializers behave a lot like initializers of the `<Module>` type.
> 
> If we were to say that module initializer is the type initializer of the type, I think it should be earlier in the chapter and the module initializer guarantees should be described as additional guarantees on top of the type initializer guarantees.

@jkotas Can you confirm which of the two directions you'd like to go? I think there could be subtleties that I wouldn't be able to appreciate yet.

Reviewer: Describing the module initializer guarantees as additional guarantees on top of the type initializer guarantees should make the spec simpler, so I would go with that.

Author: Updated. What do you think?

> modules then there is an error. In detail:~~
>
> * ~~If no member of the same kind (field or method), name, and signature exists, then
>   add this member to the output class.~~
>
> * ~~If there are duplicates and no more than one has an accessibility other than
>   **compilercontrolled**, then add them all to the output class.~~
>
> * ~~If there are duplicates and two or more have an accessibility other than
>   **compilercontrolled**, an error has occurred.~~

## Module Initializer

All modules may have a module initializer. A module initializer is defined as the type initializer (§ II.10.5.3) of the `<Module>` type (§ II.10.8).

There are no limitations on what code is permitted in a module initializer. Module initializers are permitted to run and call both managed and unmanaged code.

### Module Initialization Guarantees

In addition to the guarantees that apply to all type initializers, the CLI shall provide the following guarantees for module initializers:

1. A module initializer is executed at, or sometime before, first access to any static field or first invocation of any method defined in the module.

> 2. A module initializer shall run exactly once for any given module unless explicitly called by user code.

Reviewer: This is not needed. It is implied by the guarantees provided by the type initializers.

Author: @jkotas The type initializer guarantee would only apply to first access to members within the `<Module>` type, right? By contrast, the module initializer guarantee applies to first access to members within all types within the module.

Reviewer: Hmm, one read it that way. We can leave this in for clarity then.
